### ISSUE TITLE
Pre submission inquiry: Submit a submodule to an existing project. 

### ISSUE THREAD

**Issue by kbarnhart** [openjournals/joss#443]
Hi Joss, 

A quick question in advance of making a submission. I am one of the developers of the [landlab](https://github.com/landlab/landlab) landscape evolution modeling package. Version 1.0 of this package has already been published in an [Earth Surface Dynamics paper](https://www.earth-surf-dynam.net/5/21/2017/esurf-5-21-2017.pdf). 

I have developed a submodule for landlab that treats layered rock ([PR here](https://github.com/landlab/landlab/pull/674)) that I would like to submit as a JOSS publication. Myself and the other involved developers would like to submit only this submodule to JOSS. We have two main reasons for this: 
-  landlab itself has already been published, and the current version of landlab contains many improvements over the 1.0 version (only some of which are mine). 
- We would prefer to not move this submodule to its own repository so that it can stay within landlab for purposes of testing, documentation integration, and future maintenance.

What is the recommended way to prepare such as submission to JOSS? At the moment I've just placed the `paper.md` and `paper.bib` within my submodule directory. Should I supply the URL of the PR or the entire repository for my submission?

Looking over your existing issues it seems that #354 and #52 are related, but do not entirely answer my question. Forgive me if I missed an issue that does address it.

**Comment by jedbrown** [comment]
I think this should be supported. I think that in most cases the feature PR should be merged prior to JOSS submission because it testifies to community endorsement. If whedon doesn't want to learn about new paths, we could use a new PR for the JOSS paper in the standard location (it could be moved in a later merge commit).

The release engineering is a little tricky for larger projects that have processes like merge windows and release candidates. Ideally the JOSS archived version would contain the new feature and have similar stability to a full project release, though it surely won't always be practical to sync timelines.

I don't know if JOSS should provide guidance on what scope merits an independent submission versus a version 2 whole-project paper. It probably depends on marginal research significance, but I find it hard to articulate a simple standard.

**Comment by kbarnhart** [comment]
@jedbrown thanks for the comments. 

We hadn't merged the PR as we thought that it would be better to make sure that any JOSS-related comments and reviews were included in the initial merge of this submodule. But it would not be a problem to merge (or to make a release). I think that merging before submission makes more sense for a few reasons (in addition to the ones you provided). For example the landlab RTFD page will include this new documentation. A reviewer would also only have to clone the `master` branch instead of checking out a `developer` branch in order to verify that the code works. 

Regarding releases, we have moved towards making quarterly releases and making releases if necessary in advance of workshops, etc. 

Would it make sense for me to make a release for submission? In addition to this specific submodule, this release would probably include some other things that have been added to landlab since our last release. We could plan to make an additional release after a final post-review version. 

A follow on comment regarding what merits an independent submission vs. a whole-project paper. One of the things the landlab developers have discussed a lot is how do we give appropriate credit to the development of landlab process components or utilities (which exist as submodules). We identified JOSS as a good outlet when the authors are interested in publishing their work, but are not necessarily interested in writing a full-on science manuscript. We also identified this particular submodule as a good first submission from the landlab project to JOSS as the development team thought that it was a big enough contribution to landlab to merit its own publication. We are definitely open to feedback regarding what is substantial enough.

**Comment by labarba** [comment]
In my opinion, the release should be made after applying revisions from the review. Normal JOSS process is to make an archive in Zenodo or similar after revisions, and then the release number will be part of the Zenodo metadata. One concern about the JOSS submission being a submodule of a larger package is how to make such a Zenodo archive (do you include the whole package?). Another concern is if a second submodule wants to be submitted to JOSS (how do you add a second `paper.md` in the same repo?). So even if in principle, as @jedbrown, this "should be supported," in practice I don't know if we're prepared for it.

**Comment by labarba** [comment]
Let me ask you, could each submodule be in its own GitHub repo, then included in the big packages repo as a GitHub submodule?
https://blog.github.com/2016-02-01-working-with-submodules/
Then the JOSS submission refers to the smaller repo?

**Comment by kyleniemeyer** [comment]
This is likely something we will always have to discuss on a case-by-case basis. It seems like there are two options: 

(1) Based on @kbarnhart's description, my understanding is that this submission would be of a new version of the primary code that includes some new functionality (i.e., the submodule). If we treat this like prior new version submissions, then I think it should be merged and potentially released as a beta of the new version, pending changes after reviewer feedback.

(2) However, if the intended review is truly just of this submodule (I'm imagining a standalone application that uses the original package as a library—not sure if that is accurate here), then I think @labarba's suggestion of a separate repo using the Git submodule approach would work.

However, my impression is that this new content would be merged into the larger package, so I think the first option might be more appropriate.

**Comment by jedbrown** [comment]
Git submodules are a poor abstraction unless the submodule has its own test suite and makes its own releases -- it behaves poorly with conflict resolution in merges and with topic branch workflows driven from the parent repository.

> Another concern is if a second submodule wants to be submitted to JOSS (how do you add a second paper.md in the same repo?).

If we can't teach whedon to find paper.md in a different directory, then `paper.md` could be moved after publication to make room for other submissions. But that is disruptive centralization if we insist that the modules be merged prior to publication (which I think is a good standard).

**Comment by jedbrown** [comment]
> I'm imagining a standalone application that uses the original package as a library—not sure if that is accurate here

This is more like a plugin -- all users of the library can use this new feature to solve a class of problems that was not possible before. It makes sense to distribute with the library rather than as an aftermarket plugin for maintenance/installation convenience.

**Comment by kbarnhart** [comment]
Thanks all for your thoughtful discussion. 

I'm going to tag @mcflugen and @gregtucker here since they may have additional comments and perspective to bring in. 

To follow up on @kyleniemeyer's distinction. I would describe our intended submission as closer to option (2) but not exactly. However, I wouldn't describe it as:

> (I'm imagining a standalone application that uses the original package as a library—not sure if that is accurate here)

Perhaps a short description of landlab would also help the disucussion. landlab is a framework for building 2D models, typically of Earth surface processes. The 1.0 version of landlab provides a great deal of core functionality (such as irregular and regular 2D grids) and a number of process components that represent one type of physical process (e.g. rivers eroding rock based on a particular formulation). Each of these process components is a submodule. 

Some of these process components are very small and straightforward, and others are the major part of a PhD. The process component I've prepared and hope to submit is somewhere in between (it adds the ability to treat layered rock). So while my submodule adds new functionality to landlab, and my intention is to have just that submodule reviewed, the submodule is not really a stand alone application. 

An additional case that I think is worth considering is that landlab could have multiple submodules that represent JOSS submissions (which sounds like it could pose a problem for whedon in its current form). I know that other developers (@siccarpoint) are considering preparing JOSS submissions related to landlab. 

landlab in its present version (1.5.2) has many additions beyond the one submodule I'm hoping to submit to JOSS. These additions include internal refactoring, small process components and utilities, and contributions that have associated publications. Thus, I'd like this submission to only pertain to the new submodule I've developed. Ideally the documentation and testing that I've developed are also run as part of the standard landlab CI. 

I'm not super familiar with using git submodules (@mcflugen do you have opinions on this) but based on @jedbrown's comments I suspect that we'd prefer to go about it in a different way. 

I'm happy to be a test case in figuring out how to best handle this.

**Comment by labarba** [comment]
> If we can't teach whedon to find paper.md in a different directory, then paper.md could be moved after publication to make room for other submissions.

`whedon` searchers for `paper.md` anywhere in the repo. Therefore, if a new paper were to be submitted from the same repo, the old paper file has to be deleted or renamed. The JOSS workflow assumes that a submission is a repository.

**Comment by kbarnhart** [comment]
Would a possible approach be to add an optional field in the submission form that specifies the path to the subdirectory within the repository associated with the submission. 

Then whedon could be taught to look for certain things (e.g. the `paper.md`) only within that directory. This would allow for JOSS `paper.md`s to stay adjacent to the code they are associated with and a repository to potentially have multiple `paper. md`s.

It would also provide a way at submission to indicate that the submission is only related to the content in that subdirectory while permitting the documentation, packaging, and CI testing to stay integrated with the main package.

One barrier I'm not sure how to address is creating a DOI associated only with the subdirectory.

**Comment by jedbrown** [comment]
I think we should add a submission field for a directory under which to look for the paper. A default of empty or `/` would produce the present behavior. The alternatives complicate the submission workflow and/or create incentives that are not aligned with community research software.

To @SiccarPoint's (1) above, I (personally) think the review timeline and possible impact on complementary feature development and project maintenance are not worth the concreteness in defining what is "in" that submission. Note that significant features are often the result of a sequence of well-defined PRs.

**Comment by SiccarPoint** [comment]
I like the sound of that. And that means I'm also convinced on (1). The new field would absolutely do the job.

**Comment by labarba** [comment]
Note that changes to the submission will need to be discussed at the level of the editorial board and/or approved by the Editor-in-Chief.

**Comment by kbarnhart** [comment]
I wanted to check in on what I should do regarding my specific submission. I'm presently ready to submit, but have held off while this discussion takes its course.

**Comment by labarba** [comment]
@arfon — This thread needs EiC attention, and possible board discussion.

**Comment by arfon** [comment]
:eyes: hi folks, apologies for the slow response time on my end. This is an interesting thread, thanks for posting.

A few points of clarification:

- We definitely want to be able to support something like this (but currently don't)
- While Git submodules could work for JOSS here, I struggle to recommend them to _anyone_ as they're so confusing to work with
- Whedon **does** support multiple `paper.md` files, but not in the bot form (i.e. in review threads). When running in local/interactive mode Whedon detects all `paper.md` files and [asks the user to pick which one they would like to compile](https://github.com/openjournals/whedon/blob/ebb771b1357fe0c392ef93752dc5cfe541dff689/lib/whedon/processor.rb#L52-L62)
- Currently, Whedon only looks on the `master` or whatever the git default branch is for papers which means that pull requests have to be merged before Whedon (in bot form) can compile a paper.

What I think we should do:

- **In the (very) short term:** We can handle this submission, albeit manually. I can compile the `paper.md` file locally based on the changes in the `barnhark/make_rock_block` fork.
- **In the short term:** We should allow authors to specify a branch name (or perhaps location?) when submitting a paper to help Whedon compile the correct `paper.md`
- **In the longer term:** We should find a way for Whedon (in bot form) to handle compiling papers that are in multiple locations, in different branches, in forks etc. 

Open questions:
- Per the comment above about what exactly to archive in the final 'create an archive of the software in Zenodo step': I think the workflow we should encourage here is for authors to create a new release based on the changes in a given pull request once it has been merged (in this case https://github.com/landlab/landlab/pull/674), this will trigger Zenodo to make a new archive of the _whole repository_, including the changes. Does this sound OK to people?

**Comment by jedbrown** [comment]
Re: Zenodo archive: Yes, making a release (could be a patch or pre-release) of the whole repository sounds like the right thing. For projects with infrequent major/feature releases, this could cause some time delay between JOSS review completion and making the archive (e.g., while other components stabilize), but I think that's okay.

I think path to `paper.md` is the field we should offer in the submission. Being in the default branch ('master' below) is the signal of community acceptance that I think we should strive for. That said, the review process is iterative and large projects with merge windows make it disruptive to land every change on 'master' before the next iteration, thus a `@whedon generate pdf on branch joss-revisions` command would be handy in the review threads to save reviewers from needing to run Whedon locally. The 'joss-revisions' branch should be merged to 'master' before archival so I don't think it should be a field in the submission.

**Comment by arfon** [comment]
> `@whedon generate pdf on branch joss-revisions`

I managed to teach @whedon how to do this over the weekend. The syntax is:

```
@whedon generate pdf from branch custom-branch-name
```

**Comment by kbarnhart** [comment]
Thanks all for all the additional thoughtful comments and for teaching `@whedon` new skills!

I agree with @arfon and @jedbrown 's comments regarding what exactly gets a DOI. This will be totally fine for the landlab case. 

At the moment I've left the branch unmerged, but can merge it before submitting if need be. Let me know what makes the most sense. 

At this point I'm not totally what I need to do in order to submit. I think it is one of the following:

(optionally 0)  Merge the working branch into the `master` branch of the landlab repository

(1) Provide the information required in the submission form here (or somewhere else?) so that @arfon can handle the submission manual. 

(2) Use the submission form but put the pull request URL in the repository URL field. 

(3) Something else? 

 Again, thanks all for the thoughtful recommendations and discussion. I appreciate your enthusiasm for supporting submissions like this.

**Comment by kbarnhart** [comment]
Hi All, 
I wanted to ping again on this thread and see what I should do next. 
Thanks!

**Comment by arfon** [comment]
Hi @kbarnhart, for now I think it would be easiest if you could just merge to `master` and leave a note about this in the submission form (which I will see in the pre-review checks).

Does that sound OK?

**Comment by kbarnhart** [comment]
@arfon that sounds great. Its OK with me to close this issue unless you want to keep it open as a reminder of the eventual changes to submission structure/whedon necessary to support similar cases (and other ones in the same repository) in the future. 

thanks!

**Comment by margauxmouchene** [comment]
Hello all,

I'm re-opening this conversation because I developed a new module for Landlab and I would like to submit a paper describing it to JOSS. 

The new module functionally lives at the same level in the Landlab framework as @kbarnhart 's Lithology: the code is stored in a sub-directory of the Landlab Github repo, it is not a standalone, but a significant addition to the main code added to `master` as part of a (future) release.

@kbarnhart  mentioned the above discussion and, before submitting anything, I wanted to check with the editors (1) if it is possible for us to submit a new module that is also part of the Landlab framework (but very different from Katy's contribution), and, if so (2) what would be the procedure. In particular:

- should my new code be merged to master before submission?

- there would be two `paper.md` files in `master`: @kbarnhart 's and mine, but they would be stored in different sub-repo. Is there a way for you (and Whedon) to deal with this? Alternatively, is there a way to specify the path to the sub-directory, instead of the path to the whole Github repo?

Thanks!

**Comment by arfon** [comment]
> @kbarnhart mentioned the above discussion and, before submitting anything, I wanted to check with the editors (1) if it is possible for us to submit a new module that is also part of the Landlab framework (but very different from Katy's contribution)

Yes, this is possible.

> if so (2) what would be the procedure.

Good question :smile:

> * should my new code be merged to master before submission?

Possibly. Is it currently in a branch on the repository or in a separate fork?

> * Is there a way for you (and Whedon) to deal with this?

Yes, Whedon won't be able to deal with this automatically (i.e. some of the functionality in the review issue probably won't work) but us humans can handle this.
